### PR TITLE
zellij: add module

### LINKED ISSUE #2666: module request: zellij
### Description

[zellij](https://github.com/zellij-org/zellij) is terminal workspace written in Rust. It supports configuration via a rather large configuration file. It would be great if home-manager would contain a module for zellij

### PR DESCRIPTION
**Pull request by mainrs**
Closes #2666

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`. **Edit:** This fails on my machine with `error: attribute 'shellDryRun' missing` at `home-manager/modules/programs/bash.nix:12:9`. I didn't touch that module.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/programs/zellij.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        type = yamlFormat.type;
+        default = { };
+        defaultText = "{}";
+        example = literalExpression ''
+          {
+            theme = "custom";
+            themes.custom.fg = 5;
+          }
+        '';
+        description = "The zellij configuration.";
```
Comment (berbiche) [inline modules/programs/zellij.nix:33]:
```suggestion
        description = ''
          Configuration written to
          <filename>$XDG_CONFIG_HOME/zellij/config.toml</filename>.
          </para><para>
          See <link xlink:href="https://www.zellij.sh/book/configuration.html" /> for the full list
          of options.
        '';
```

File: modules/programs/zellij.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    };
+  };
+
+  meta = {
+    description = "A terminal workspace with batteries included.";
+    homepage = "https://zellij.dev/";
+    maintainers = [ maintainers.mainrs ];
+    platforms = platforms.unix;
+    license = licenses.mit;
+  };
```
Comment (berbiche) [inline modules/programs/zellij.nix:51]:
```suggestion
  meta.maintainers = [ maintainers.mainrs ];
```
AFAIK the module system does not use/have other meta attributes.

**Review by berbiche — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: modules/programs/zellij.nix (hunk 2 above)
Comment (mainrs) [reply modules/programs/zellij.nix:51]: I'll change this in one of my other PRs for `eww` as well! Thanks for letting me know!

**Comment by happysalada** [comment]
@berbiche friendly ping (no worries if you're busy and no rush.), I'm interested by this.

File: modules/programs/zellij.nix (hunk 1 above)
Comment (berbiche) [reply modules/programs/zellij.nix]: Yup, the description is contained within a `<para>graph` and is automatically closed by one.
Adding new paragraphs is equivalent to adding a linebreak `<br>`.

File: modules/programs/zellij.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+      settings = mkOption {
+        type = yamlFormat.type;
+        default = { };
+        defaultText = "{}";
+        example = literalExpression ''
+          {
+            theme = "custom";
+            themes.custom.fg = 5;
+          }
```
[the remark this answers is not in the document]
Comment (mainrs) [reply modules/programs/zellij.nix:31]: I took a look at the code and it uses the normal HashMap implementation which doesn’t make ordering guarantees. So I’d say no, it doesn’t matter.

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -428,3 +428,5 @@ Makefile                                              @thiagokokada
 /modules/programs/ion.nix                             @jo1gi
 
 /modules/services/plex-mpv-shim.nix                   @starcraft66
+
+/modules/programs/zellij.nix                          @mainrs
```
Comment (sumnerevans) [inline .github/CODEOWNERS:432]: This should go up with the other `/modules/programs`. Not sure why ion is down here, but it shouldn't be either.

File: modules/programs/zellij.nix [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        example = literalExpression "pkgs.zellij";
+        description = ''
+          The zellij package to install.
+        '';
+      };
+
+      settings = mkOption {
+        type = yamlFormat.type;
+        default = { };
+        defaultText = "{}";
```
Comment (sumnerevans) [inline modules/programs/zellij.nix:26]:
```suggestion
        defaultText = literalExpression "{ }";
```

**Review by sumnerevans — CHANGES_REQUESTED** [review]

File: .github/CODEOWNERS (hunk 4 above)
Comment (mainrs) [reply .github/CODEOWNERS:432]: I don't think I read somewhere that it had to be alphabetically sorted. Might be worth doing that once in a separate PR and updating `CONTRIBUTING`. I might have missed it though.

**Review by berbiche — APPROVED** [review]
LGTM!

File: modules/programs/zellij.nix [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    home.packages = [ cfg.package ];
+    xdg.configFile."zellij/config.yaml" = mkIf (cfg.settings != { }) {
+      source = yamlFormat.generate "zellij.yaml" cfg.settings;
+    };
+  };
+
+  meta.maintainers = [ maintainers.mainrs ];
```
Comment (berbiche) [inline modules/programs/zellij.nix:51]:
```suggestion
  meta.maintainers = [ hm.maintainers.mainrs ];
```

I missed this.

**Comment by rycee** [comment]
Merged to master now. Thanks!

**Comment by happysalada** [comment]
in the zellij docs, there are ways to unbind certain keybindings

the conf should be 
```
keybinds:
    unbind: [ Ctrl: 'p']
```

for the love of everything that is holy, I couldn't find the equivalent nix expression that would yield the corresponding yaml.

this is what I tried
```
keybinds = {
   unbind = [ "Ctrl: 'p'"];
};
```
as soon as I try to uncomment inside the square brackets, it doesn't preserve them.

[reference](https://zellij.dev/documentation/keybindings.html)

**Comment by berbiche** [comment]
@happysalada 
> in the zellij docs, there are ways to unbind certain keybindings
> 
> the conf should be 
> ```
> keybinds:
>     unbind: [ Ctrl: 'p']
> ```

YAML allows using JSON notation for inline arrays (instead of using `-`).
The content of `unbind` is actually an array of one element, an object with key `Ctrl` and value `"p"`.

I think this should be what you are looking for:
``` nix
keybinds.unbind = [ { "Ctrl" = "p"; } ];
```

**Comment by happysalada** [comment]
The formatting is not the same, which got me confused at first, but it actually works!
Thanks a lot for this!

**Pull request merged**